chainer/training/extensions/parameter_statistics.py

```diff
@@ -0,0 +1,187 @@
+            or CuPy functions.
+
+    Returns:
+        dict: Mappings from functions keys to statistic values.
+    """
+    stats = {}
+    for f in functions:
+        try:
+            # nan if x.size == 0 and f in ('mean', 'std')
+            stats[f] = getattr(x, f, lambda: float('NaN'))()
+        except ValueError:  # x.size == 0 and f in ('min, 'max')
+            stats[f] = float('NaN')
+
+    return stats
+
+
+def _percentiles(x, sigmas):
+    """Compute percentiles for the given array.
+
+    Args:
+        x (array): Target array for which percentiles are computed.
+        sigmas (iterable): Percentile sigma values.
+
+    Returns:
+        array: List of percentiles. The list has the same length as the given
+            ``sigma``.
+    """
+    def _percentiles_cpu(_x):
+        try:
+            return numpy.percentile(_x, sigmas).astype(_x.dtype)
+        except IndexError:  # _x.size == 0
+            return numpy.array((float('NaN'),) * len(sigmas))
+
+    if cuda.available and isinstance(x, cuda.ndarray):
+        # cuda.cupy.percentile() is not implemented
+        return cuda.to_gpu(_percentiles_cpu(cuda.to_cpu(x)))
+
+    return _percentiles_cpu(x)
+
+
+def _zeros(x):
+    """Count the number of zeros in the given array.
+
+    Args:
+        x (array): Target array for which sparsity is computed.
+
+    Returns:
+        int: Number of zeros.
+    """
+    if x.ndim == 0:
+        raise ValueError('Cannot count zeros for shape {}'.format(x.shape))
+
+    return x.size - cuda.get_array_module(x).count_nonzero(x)
+
+
+class ParameterStatistics(extension.Extension):
+    """Trainer extension to report parameter statistics.
+
+    The statistics are collected for a given `~chainer.Link` or an iterable of
+    `~chainer.Link`s. If a link contains child links, the statistics are
+    aggregated over all its children.
+
+    Statistics that can be collected and reporter using the current scope are
+    minimum and maximum values, means, standard deviations, percentiles and
+    zero counts.
+
+    Args:
+        links (~chainer.Link or iterable of ~chainer.Link): Links containing
+            the parameters to observe. The link is expected to have a ``name``
+            attribute which is used as a part of a key in the report.
+        trigger: Trigger that decides when to aggregate the results and report
+            the values.
+        count_zeros (bool): If ``True``, count the number of zero elements and
+            include those statistics in the report. Else, do not compute the
+            number of zero elements.
+        prefix (str): Prefix to prepend to the report keys.
+        targets (iterable of tuples): Target parameters and their attributes
+            to consider in this extension.
+    """
+    default_name = 'parameter_statistics'
+    priority = extension.PRIORITY_WRITER
+
+    def __init__(self, links, trigger=(1, 'epoch'), count_zeros=False,
+                 prefix=None, targets=(('W', 'data'),
```

Comment:
How about using `params` to check all parameters of a link for default behavior?